hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/SourceBuilder.java

```diff
@@ -26,46 +29,54 @@
 import com.hazelcast.util.Preconditions;
 
 import javax.annotation.Nonnull;
+import java.util.List;
 
 import static com.hazelcast.jet.core.processor.SourceProcessors.convenientSourceP;
 import static com.hazelcast.jet.core.processor.SourceProcessors.convenientTimestampedSourceP;
 import static com.hazelcast.util.Preconditions.checkPositive;
 
 /**
- * Top-level class for Jet source builders. Refer to the builder
- * factory methods listed below.
+ * Top-level class for Jet custom source builders. It is associated
+ * with a <em>context object</em> that holds any necessary state and
+ * resources you need to make your source work.
+ * <p>
+ * For further details refer to the factory methods:
+ * <ul>
+ *     <li>{@link #batch(String, FunctionEx)}
+ *     <li>{@link #timestampedStream(String, FunctionEx)}
+ *     <li>{@link #stream(String, FunctionEx)}
+ * </ul>
  *
- * @see #batch(String, FunctionEx)
- * @see #timestampedStream(String, FunctionEx)
- * @see #stream(String, FunctionEx)
- *
- * @param <S> type of the source state object
+ * @param <C> type of the context object
  */
-public final class SourceBuilder<S> {
-    // the m-prefix in field names is there to allow unqualified access from
-    // inner classes while not clashing with parameter names. Parameter names
-    // are public API.
-    private final String mName;
-    private final FunctionEx<? super Context, ? extends S> mCreateFn;
-    private ConsumerEx<? super S> mDestroyFn = ConsumerEx.noop();
-    private int mPreferredLocalParallelism;
+public final class SourceBuilder<C> {
+    private final String name;
+    private final FunctionEx<? super Context, ? extends C> createFn;
+    private FunctionEx<? super C, Object> createSnapshotFn;
+    private BiConsumerEx<? super C, ? super List<Object>> restoreSnapshotFn;
+    private ConsumerEx<? super C> destroyFn = ConsumerEx.noop();
+    private int preferredLocalParallelism;
 
     /**
      * The buffer object that the {@code fillBufferFn} gets on each call. Used
      * in sources that emit items without a timestamp.
      *
-     * @param <T> type of the buffer item
+     * @param <T> type of the emitted item
      */
     public interface SourceBuffer<T> {
+
         /**
          * Returns the number of items the buffer holds.
          */
         int size();
 
         /**
-         * Closes the buffer, signaling that all items have been emitted.
+         * Closes the buffer, signaling that all items have been emitted. Only
+         * {@linkplain #batch} sources are allowed to call this method.
+         *
+         * @throws JetException if the source is a streaming source
          */
-        void close();
+        void close() throws JetException;
```

Comment:
this is a breaking change

Reply:
`JetException` is unchecked, is it really a breaking change?